module/scheduler_components/deterministic_lock_manager.cpp

```diff
@@ -196,6 +196,92 @@ DeterministicLockManager::ReleaseLocks(const Transaction& txn) {
   return ready_txns;
 }
 
+VerifyMasterResult
+DeterministicLockManager::VerifyMaster(const Transaction& txn) {
+  auto keys = ExtractKeys(txn);
+  if (keys.empty()) {
+    // None of the keys in this txn is in this partition
+    return VerifyMasterResult::Valid;
+  }
+
+  auto txn_master_metadata = txn.internal().master_metadata();
+  if (txn_master_metadata.empty()) { // allow this to not be set for testing
+    LOG(WARNING) << "Master metadata empty: txn id " << txn.internal().id();
+    return VerifyMasterResult::Valid;
+  }
+
+  // keeps track of counters that will need to be waited on
+  std::unordered_map<Key, int32_t> ahead_counters;
+  for (auto pair : keys) {
+    auto key = pair.first;
+
+    Record record;
+    bool found = storage_->Read(key, record);
```

Comment:
Hmm I thought it would be safe from reading the reference, am I misunderstanding it?
> All member functions (including copy constructor and copy assignment) can be called by multiple threads on different instances of shared_ptr without additional synchronization even if these instances are copies and share ownership of the same object

Reply:
This is true for the shared_ptr object, not the object that the pointer is pointing to.